e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -0,0 +1,230 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package portforwarder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest/retry"
+	"kpt.dev/configsync/e2e/nomostest/testkubeclient"
+	"kpt.dev/configsync/e2e/nomostest/testlogger"
+	"kpt.dev/configsync/e2e/nomostest/testwatcher"
+	"kpt.dev/configsync/pkg/kinds"
+)
+
+// PortForwarder creates a port forwarding from localhost to a Deployment in the
+// cluster. Assumes the Deployment has a single Pod, and keeps the port forward
+// open if the Pod is deleted.
+type PortForwarder struct {
+	// kubeClient is used for list/get operations on k8s objects in the cluster
+	kubeClient *testkubeclient.KubeClient
+	// watcher is used to watch k8s objects in the cluster
+	watcher *testwatcher.Watcher
+	// logger is used for logging
+	logger *testlogger.TestLogger
+	// retryTimeout is used to determine the retry duration
+	retryTimeout time.Duration
+	// ctx should be a cancellable Context. Once ctx is cancelled, the port forward
+	// will be closed and will no longer be recreated.
+	ctx context.Context
+	// kubeConfigPath is the path to the kube config file for the cluster
+	kubeConfigPath string
+	// ns is the Namespace of the Deployment/Pod
+	ns string
+	// deployment is the name of the Deployment. Assumes a single Pod replica.
+	deployment string
+	// port is the port on the Pod to forward to
+	port string
+	// setFn is a callback which will be invoked whenever the Pod changes and the
+	// port-forward is recreated.
+	setFn func(int, string)
+	// mux is a mutex to synchronize getting/setting the port-forward value
+	mux sync.Mutex
+	// localPort is the local port which forwards to the Pod
+	localPort int
+	// pod is the name of the current Pod that is being forwarded to
+	pod string
+}
+
+// PortForwardOpt is an optional parameter for PortForwarder
+type PortForwardOpt func(pf *PortForwarder)
+
+// WithPortForwardSetFn registers a callback that will be invoked whenever the
+// deployment's underlying pod changes and the port-forward is recreated.
+func WithPortForwardSetFn(setFn func(int, string)) PortForwardOpt {
```

Comment:
"PortForward" is redundant with the package now. This can just be WithSetFn. But fwiw, I usually prefer "func" over "fn" for these. And "setFunc" isn't very self-descriptive... if it's called when the forwarding is set up, maybe "setupFunc" or "initFunc"?

Reply:
renamed